### ISSUE TITLE
NYC doesn't work correctly with local node_modules

### ISSUE THREAD

**Issue by hershmire** [istanbuljs/nyc#898]
## Expected Behavior

NYC to not exclude local node_modules (i.e. `src/node_modules/...` OR `src/somedirectory/node_modules/**`) BUT to exclude root node_modules. We have a lot of Node applications at the company I work for that use the pattern of creating local node_modules as it helps modularize our code at an application/page-level and provides a nice way to require files instead of doing long relative paths to everything. Below is an example file structure many of our apps could see:

```
app/
  |- node_modules/ (root modules and we expect to exclude)
  |- src/
  |  |- node_modules/ (local to the app)
  |  |  |- local-module-1/
  |  |  |  |- test-node/
  |  |  |  |  |- file-test.js
  |  |  |  |- package.json
  |  |  |  |- file.js
  |  |- middleware
  |  |- pages/
  |  |  |- page-1/
  |  |  |  |- test-node/
  |  |  |  |  |- file-test.js
  |  |  |- file.js
  |  |  |  |- node_modules/ (local to page 1 of application)
  |  |  |  |  |- page-local-module-1/
  |  |  |  |  |  |- test-node/
  |  |  |  |  |  |  |- file-test.js
  |  |  |  |  |  |- package.json
  |  |  |  |  |  |- file.js
  |  |  |- page-2/
  ...
```

## Observed Behavior

Since by default, NYC adds the glob `**/node_modules/**` it will exclude all node_modules, which in our case we don't want. To get around this problem, you have a negateExcludes section where we supply `!src/**/node_modules/**` but that doesn't quite work right as it will force to include our test files which live next to the files that are being tested since negateExclude is evaluated after the exclude rules.

Additionally, even with the `all: true` flag set in package.json, NYC doesn't appear to be picking up all files as I thought it would; especially the ones that I purposely left out of having tests point to – i.e. https://github.com/hershmire/nyc-test/blob/master/src/pages/page-1/components/Name.jsx & https://github.com/hershmire/nyc-test/tree/master/src/pages/page-no-tests-1.

It would be nice to not exclude all node_modules by default but just the root ones for starters. OR at least provide a easier way to turn that off. It gets quite complicated when you also have to take into account babel-plugin-istanbul and understand how to set both up properly.

## Bonus Points! Code (or Repository) that Reproduces Issue

I created a quick demo app that shows what I'm running into here:

https://github.com/hershmire/nyc-test

After running `yarn install`, run `yarn coverage`

As you can see from the results, my test files under any `test-node` folders are picked up which they should be excluded.

![screen shot 2018-08-06 at 13 33 33](https://user-images.githubusercontent.com/757524/43740748-05c96f44-9981-11e8-8adf-533a579d39fa.png)

## Forensic Information

**Operating System:** the operating system you observed the issue on.
MacOS 10.13.6

**Environment Information:** information about your project's environment, see instructions below:

1. run the following script:

sh -c 'node --version; npm --version; npm ls' > output.txt

https://github.com/hershmire/nyc-test/blob/master/output.txt

**Comment by hershmire** [comment]
Anyone have time to take a look at this and provide some thoughts?

**Comment by leslc** [comment]
This is affecting our ability to run unit tests.  Any solution, even a hacky workaround, would be appreciated.  Or where the fix could be made with a PR.

**Comment by coreyfarrell** [comment]
This is not a simple issue, we cannot just stop ignoring `**/node_modules/**`, this is needed to support monorepo's (for example lerna repos).  I'll have to look into a way to make this easier to override but the default is what it needs to be.

Sorry I don't have any suggestions for work-arounds, you are doing things in a way I've never thought of.

**Comment by hershmire** [comment]
@coreyfarrell Thanks for the reply! Let me know if there is anything we can do to help you out with this.

**Comment by leslc** [comment]
Thanks @coreyfarrell

Question on the glob.sync to return all file names.  I see `{ignore: this.exclude.exclude}` is passed to glob.sync.  Does this mean the "negatedExclude" is skipped because it's not called from `this.exclude.shouldInstrument()`?
https://github.com/istanbuljs/nyc/blob/master/index.js#L249

**Comment by leslc** [comment]
Nope, have to take that back. glob.sync doesn't support negation in the ignore.

**Comment by jweidler** [comment]
Unfortunately there are problems with the misunderstanding of how `node_modules` are not exclusively third-party modules on every corner. This is broken and won't be fixed. So we've moved away from using local `node_modules` to defining local dependencies via `package.json` and have npm create symlinks. This approach requires some wiring upfront but once it's set up it feels much nicer than having `node_modules` all over the place.

Here in detail how we solved it:
https://github.com/istanbuljs/nyc/issues/833#issuecomment-406587243

**Comment by hershmire** [comment]
> Unfortunately there are problems with the misunderstanding of how node_modules are not exclusively third-party modules on every corner.

Yea, it seems there is a big misunderstanding of that fact indeed. Thanks for passing along your solution to NYC's limitation but unfortunately, we have too many applications that are using local node_modules and I'm concerned about the risks when building and deploying out to different environments. Until we can get some additional dialogue with the NYC owners for a proper solution, we might just fork the repo and handle the fix internally that way. 

I'm happy to contribute to NYC's codebase for a fix but first need more feedback from the owners/contributors.

**Comment by coreyfarrell** [comment]
> Unfortunately there are problems with the misunderstanding of how node_modules are not exclusively third-party modules on every corner.

I don't agree with this assessment.  node_modules does have a specific purpose, to install external modules.  It's possible to abuse it as you have, but node_modules is meant for stuff installed by npm - normally stuff that has it's own testing.  I'm not against making it possible/easier for you to use NYC but I have to prioritize the normal way of working.

I'm open to making a boolean option to prevent test-exclude from forcing exclude of `**/node_modules/**`.  So you would configure NYC to not force it, then configure your own excludes which has `node_modules/**` instead of `**/node_modules/**`.

**Comment by leslc** [comment]
thanks @hershmire, this is exactly what we need. @coreyfarrell, thoughts on merging these PRs?

**Comment by coreyfarrell** [comment]
@leslc I have to ask for patience.  I'm currently dealing with release related issues which need to be addressed first, I have limited time to work on NYC.

**Comment by hershmire** [comment]
@coreyfarrell Understandable that you're busy. Any idea when you'll be able to have time to look at the three PRs referenced above?

**Comment by theqabalist** [comment]
This seems to also match our use case at our job as well.  So I would also appreciate any attention this matter could receive.
